src/hook.ts

```diff
@@ -37,14 +37,16 @@ export type StateSelectorHook<T> = {
  */
 export type StoreHook<T extends AnyState> = StateSelectorHook<T> & StoreAPI<T>;
 
-function bindHook<T extends AnyState>(
+const bindHook = <T extends AnyState>(
   store: StoreAPI<T>
-): StateSelectorHook<T> {
-  const useSelector = (
+): StateSelectorHook<T> => {
+  const useSelectValue = (selector: StateSelector<any> = (state) => state) =>
+    selector(store.get());
+  const useSelectMemoizedValue = (
```

Comment:
See [useSyncExternalStoreWithSelector](https://github.com/facebook/react/blob/c5b9375767e2c4102d7e5559d383523736f1c902/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) for a more "robust" implementation of sync with selectors. I haven't really looked at all the detains in what it does, it was just something I found and thought might interest you.

Reply:
> I haven't really looked at all the detains in what it does, it was just something I found and thought might interest you.

Cheers for having a look into it. ❤️ 

I looked into this before publishing 1.0.0 (zustand uses it), and a bit more now. It's quite neat, but it's [a lot more code](https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) to cover [concerns](https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js#L50-L53) that Tyin doesn't have (our stores are already shared across all instances/concurrent copies—they are global), it's also (unfortunately) an external dependency.

As for why `useSelectMemoizedValue` is here: It's _only_ for clientside rendering. For SSR (when `typeof window === "undefined"`), there is no re-rendering (by definition). So instead of going through the trouble of setting up a `ref` and calling `useSyncExternalStore`, we can confidently just pull out whatever value is in the store. This is the "effective code path" that `useSyncExternalStore(WithSelector)` will take regardless, but I figure I might as well bypass it entirely.

By the way: The way you use Tyin (and zustand) with SSR is with a HOC or similar, so that you can "preload" values into the stores before you render the page. It's a bit scary since the stores are global, and the server will re-use whatever state is left in them for the next render. The way you solve that is simple though: never preload anything conditionally.

Reply:
To recap what I said (since the comments are gone).

- I had a look at `useSyncExternalStoreWithSelector` before releasing 1.0.0, and again now. 👍 
- It mostly [covers concerns that we don't have (stores are global, so instance sharing is by design)](https://github.com/facebook/react/blob/c5b9375767e2c4102d7e5559d383523736f1c902/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js#L50-L53)
- Server-side rendering can just pull the value directly from the store (no need to go through the useSyncExternalStore) even (server-side doesn't re-render, and the values never change after preloading\*)

\* Using global stores with SSR is a feels precarious since the stores are not (automatically) cleared out between requests (they are global and, thus, initialized on startup). The way you solve this is by never _conditionally_ preloading anything into the stores, that way, the data is never stale.

Reply:
Great! I haven't done much SSR, but it makes sense. 

I guess if `selector` **could be**  `undefined`, the client side code would essentially work as the SSR code? Because `equals` is only useful if the `selector` is defined? If I understand the code correctly.  E.g. 

```ts
return useSyncExternalStore(store.subscribe, select ?? store.get, select ?? store.get);
```